Rx/v2/src/rxcpp/operators/rx-throttle.hpp

```diff
@@ -0,0 +1,271 @@
+        typedef rxu::decay_t<Subscriber> dest_type;
+        typedef observer<T, this_type> observer_type;
+
+        struct throttle_subscriber_values : public throttle_values
+        {
+            throttle_subscriber_values(composite_subscription cs, dest_type d, throttle_values v, coordinator_type c)
+                : throttle_values(v)
+                , cs(std::move(cs))
+                , dest(std::move(d))
+                , coordinator(std::move(c))
+                , worker(coordinator.get_worker())
+                , throttled(false)
+            {
+            }
+
+            composite_subscription cs;
+            dest_type dest;
+            coordinator_type coordinator;
+            rxsc::worker worker;
+            mutable bool throttled;
+        };
+        typedef std::shared_ptr<throttle_subscriber_values> state_type;
+        state_type state;
+
+        throttle_observer(composite_subscription cs, dest_type d, throttle_values v, coordinator_type c)
+            : state(std::make_shared<throttle_subscriber_values>(throttle_subscriber_values(std::move(cs), std::move(d), v, std::move(c))))
+        {
+            auto localState = state;
+
+            auto disposer = [=](const rxsc::schedulable&){
+                localState->cs.unsubscribe();
+                localState->dest.unsubscribe();
+                localState->worker.unsubscribe();
+            };
+            auto selectedDisposer = on_exception(
+                [&](){ return localState->coordinator.act(disposer); },
+                localState->dest);
+            if (selectedDisposer.empty()) {
+                return;
+            }
+
+            localState->dest.add([=](){
+                localState->worker.schedule(selectedDisposer.get());
+            });
+            localState->cs.add([=](){
+                localState->worker.schedule(selectedDisposer.get());
+            });
+        }
+
+        static std::function<void(const rxsc::schedulable&)> reset_throttle(state_type state) {
+            auto reset = [state](const rxsc::schedulable&) {
+                state->throttled = false;
+            };
+
+            auto selectedReset = on_exception(
+                    [&](){ return state->coordinator.act(reset); },
+                    state->dest);
+            if (selectedReset.empty()) {
+                return std::function<void(const rxsc::schedulable&)>();
+            }
+
+            return std::function<void(const rxsc::schedulable&)>(selectedReset.get());
+        }
+
+        void on_next(T v) const {
+            auto localState = state;
+
+            const auto tp = localState->worker.now().time_since_epoch();
+            std::cout << "on_next(" << v << ") at " << tp.count() / 1000000 << " throttled: " << (localState->throttled ? "true" : "false") << std::endl;
+
+            if (!localState->throttled) {
+                localState->throttled = true;
+
+                state->dest.on_next(v);
+
```

Comment:
the call to on_next is not necessarily on the same worker. the following needs to be scheduled on the worker.

```cpp
            if (!localState->throttled) {
                localState->throttled = true;

                state->dest.on_next(v);
            }
```

It would be more efficient to check `throttled` first and then only schedule `state->dest.on_next(v);` when `throttled == false`. To do that throttled would need to be `std::atomic<bool>`.